Python/pywarpx/diagnostics/data/BaseReader.py

```diff
@@ -0,0 +1,185 @@
+        self.fname = None
+
+        self.has_species = None
+
+    def get_data_path(self):
+        """
+        Returns
+        -------
+        A string with the path to the current diagnostic (assumed unique).
+        """
+
+        if not os.path.isdir(self.red_diags_dir):
+            raise IOError(f'The reduced diagnostic directory {self.red_diags_dir} does not exist.\n'
+                          'Did you set the proper path to the reduced diagnostic directory?\n'
+                          'Did the simulation already run?')
+
+        data_file_path = os.path.join(
+            self.red_diags_dir,
+            self.fname
+        )
+
+        if not os.path.isfile(data_file_path):
+            raise IOError(f'The file {data_file_path} does not exist.\n'
+                          'Did you type the correct filename?')
+
+        return data_file_path
+
+    def get_ncols(self):
+        ''' Returns the number of columns of the reduced diagnostic.'''
+        data_file_path = self.get_data_path()
+        header=pd.read_csv(data_file_path, header=0, sep=self.separator, nrows=0)
+        ncols = len(header.columns)
+        return ncols
+
+    def get_nspecies(self, columns_speciesless=None, columns_per_species=None):
+        '''Returns the number of species involved in the diagnostic.'''
+        # deduces number of species from number of columns
+        if self.has_species == True:
+            ncols = self.get_ncols()
+            # remove <columns_speciesless> columns and divide to avoid counting multiple entries
+            nspecies = int((ncols - columns_speciesless)/columns_per_species)
+            return nspecies
+        else:
+            return 0
+
+    def get_species_names(self, string=None, initial_columns=None, columns_per_species=None):
+        '''Returns the names of the species involved in the diagnostic.'''
+        # deduces names of species from the header by considering only the relevant entries
+        # finds the string by locating the units
+        if self.has_species == True:
+            species_names = []
+            data_file_path = self.get_data_path()
+            nspecies = self.get_nspecies()
+
+            header = pd.read_csv(data_file_path, header=0, sep=self.separator, nrows=0)
+            colnames = header.columns.values
+            for i in range(nspecies):
+                result=re.search('](.+?)'+string,colnames[initial_columns+i*columns_per_species])
+                species_names.append(result.group(1))
+            return species_names
+        else:
+            raise ValueError('The current diagnostics does not involve any species!')
+
+    def load_data(self):
+        data_file_path = self.get_data_path()
+        df = pd.read_csv(data_file_path, sep=self.separator, header=0)
+        return df
+
+    def get_steps(self):
+        data_file_path = self.get_data_path()
+        df = pd.read_csv(data_file_path, sep=self.separator, usecols=['#[0]step()'])
```

Comment:
We don't need to do this too early, but technically we can and should cache these reads on first access to any of these methods to avoid repeated filesystem access.